File: amazon/93_restore_IP_adresses.go
package main

import (
	"fmt"
	"strconv"
)

func restoreIpAddresses(s string) []string {
	var result []string
	backtrack(s,0,0,"",&result)
	return result
}

func backtrack(s string, start, dots int, path string, result* []string){
	if dots == 4 && start == len(s){
		//remove traillin dot before adding to the result
		*result = append(*result, path[:len(path)-1])
		return
	}
	if dots > 4 || start >= len(s){
		return
	}
	for length :=1; length <= 3; length++{
		if start+length > len(s){
			break
		}
		segment := s[start : start+length]
		if isValid(segment){
			backtrack(s, start+length, dots+1, path+segment+".",result)
		}
	}
}

func isValid(segment string)bool{
	if len(segment) > 3 || (segment[0]=='0'&&len(segment)>1){
		return false
	}
	val, err := strconv.Atoi(segment)
	if err != nil{
		return false
	}
	return val >= 0 && val <= 255
}

File: amazon/93_restore_IP_adresses.scala

import scala.collection.mutable.ListBuffer

object Solution{
    def restoreIpAddresses(s: String): List[String] = {
        val result = ListBuffer[String]()

        def isValid(segment: String): Boolean = {
            if(segment.isEmpty || segment.length > 3) false
            else if(segment.startsWith("0") && segment.length > 1) false
            else{
                val num = segment.toInt
                num >= 0 && num <= 255
            }
        }

        def backtrack(start: Int, dots: Int, path: List[String]): Unit = {
            if(dots == 4 && start == s.length){
                result += path.reverse.mkString(".")
                return
            }
            if (dots > 4 || start >= s.length) return

            for(len <- 1 to 3 if start + len <= s.length){
                val segment = s.substring(start, start + len)
                if (isValid(segment)){
                    backtrack(start + len, dots + 1, segment :: path)
                }
            }
        }
        backtrack(0, 0, Nil)
        result.toList
    }
}